onnxruntime/core/graph/contrib_ops/internal_nhwc_onnx_opset.cc

```diff
@@ -115,10 +136,10 @@ void OpSet_Internal_NHWC_ONNX::ForEachSchema(const std::function<void(ONNX_NAMES
   REGISTER_NHWC_SCHEMA(fn, SpaceToDepth, 1);
   REGISTER_NHWC_SCHEMA(fn, SpaceToDepth, 13);
 
-#if defined(USE_QNN)
-  REGISTER_NHWC_SCHEMA(fn, Resize, 11);
-  REGISTER_NHWC_SCHEMA(fn, Resize, 13);
-#endif
+  // Used by QNN EP.
+  REGISTER_RESIZE_NHWC_SCHEMA(fn, Resize, 11);
+  REGISTER_RESIZE_NHWC_SCHEMA(fn, Resize, 13);
+  REGISTER_RESIZE_NHWC_SCHEMA(fn, Resize, 18);
```

Comment:
I don't think these need to be guarded by `#ifdef USE_QNN`, but would like confirmation.

Reply:
Additionally the setup is currently inconsistent between transpose_optimizer.cc and optimizer_api_impl.cc. 

We disable HandleResize if CUDA or ROCm are enabled to prevent changing the layout of the input to the Resize node in the generic transpose optimizer (which runs as a level 1 optimizer as well as being used by layout transformation). QNN is not included in that list. That means the transpose optimizer could push a Transpose through a Resize during level 1 optimization, which happens prior to any EP specific layout transform. 

https://github.com/microsoft/onnxruntime/blob/d76cf374c48f1849193cc958e14f1851a723cb20/onnxruntime/core/optimizer/transpose_optimizer/transpose_optimizer.cc#L973-L974

We special case Resize for CUDA, ROCm and QNN in the ORT specific logic in optimizer_api_impl.cc that is used to do layout transformation.

https://github.com/microsoft/onnxruntime/blob/d76cf374c48f1849193cc958e14f1851a723cb20/onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc#L876-L881

We can make the ORT specific logic smarter by treating Resize as layout sensitive on a per-EP basis. 

The correct thing to do in the generic transpose_optimizer.cc code is less clear. The `#ifdef` based on the EPs included in the build is not great as it assumes any special cased EP will a) be enabled at runtime, and b) will take the Resize node. If neither of those turn out to be true a Transpose potentially gets unnecessarily blocked on the Resize.

Reply:
I briefly considered treating Resize on a per-EP basis by potentially introducing a virtual `GetLayoutSensitiveOps` function on `IExecutionProvider`. This could be combined with the result of `GetORTLayoutSensitiveOps()` in `TransformLayoutForEP`.

https://github.com/microsoft/onnxruntime/blob/d76cf374c48f1849193cc958e14f1851a723cb20/onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc#L909-L913

In my naive understanding, this would obviate the need to include special `#ifdef` conditions for QNN, CUDA, and ROCM in optimizer_api_impl.cc. However, I'm not so sure how to use this in `HandleResize`.

I'm not sure this is the right direction, but I thought I would share.

Reply:
Also, as you mentioned, it would be worth investing if it's just easier for QNN to always expect Resize to be NCHW, and apply the necessary transformations within the EP itself. Hmmm.